src/main/java/net/neoforged/neoforge/transfer/handlers/wrappers/fluids/BlockFluidHandler.java

```diff
@@ -0,0 +1,197 @@
+
+    @Override
+    public int extract(FluidResource resource, int amount, TransactionContext transaction) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount)) return 0;
+        //We already are handling an insert this callback, we can't do more until closing.
+        // Something like vaporizing needs to be deferred until commit, but we don't have a
+        // system that allows us to queue callbacks as first-in-first-out.
+        if (!fluidStack.isEmpty()) return 0;
+
+        BlockState state = level.getBlockState(pos);
+        FluidState fluidState = level.getFluidState(pos);
+        if (amount < FluidType.BUCKET_VOLUME || resource.isEmpty() || !resource.equals(fluidState.getType().getDefaultResource()))
+            return 0;
+
+        if (!(state.getBlock() instanceof BucketPickup pickupHandler)) return 0;
+
+        updateSnapshots(transaction);
+
+        ItemStack stack = pickupHandler.pickupBlock(player, level, pos, state);
+        if (!stack.isEmpty()) {
+            if (stack.getItem() instanceof BucketItem bucket && !resource.equals(bucket.content.getDefaultResource())) {
+                LOGGER.error("Fluid removed without successfully being picked up. Fluid {} at {} in {} matched requested type, but after performing pickup was {}.",
+                        BuiltInRegistries.FLUID.getKey(fluidState.getType()), pos, level.dimension().location(), BuiltInRegistries.FLUID.getKey(bucket.content));
+                return 0;
+            }
+            return FluidType.BUCKET_VOLUME;
+        }
+        return 0;
+    }
+
+    @Override
+    protected BlockFluidHandler.Snapshot createSnapshot() {
+        return new Snapshot(level.getBlockState(pos), fluidStack);
+    }
+
+    @Override
+    protected void revertToSnapshot(BlockFluidHandler.Snapshot snapshot) {
+        level.setBlock(pos, snapshot.state, 0);
+        fluidStack = FluidStack.EMPTY;
+    }
+
+    @Override
+    protected void onCommit(BlockFluidHandler.Snapshot originalState) {
+        // if it should be vaporized we can skip the rest of the logic
+        if (fluidStack.getFluidType().isVaporizedOnPlacement(level, pos, fluidStack)) {
+            fluidStack.getFluidType().onVaporize(player, level, pos, fluidStack);
+            return;
+        }
+        BlockState state = level.getBlockState(pos);
+        if (originalState.state == state) return;
+
+        //Set the block back to its original momentarily
+        level.setBlock(pos, originalState.state, 0);
+        //check to see if the original was a valid container of liquid
+        boolean isValidLiquidContainer = originalState.state.getBlock() instanceof LiquidBlockContainer container && container.canPlaceLiquid(player, level, pos, originalState.state, fluidStack.getFluid());
+
+        if (!isValidLiquidContainer) {
+            //Apply the normal destruction if the block wasn't a container when placing a bucket of fluid onto the block resulting in that block to be broken.
```

Comment:
Add to the comment that this should be validated as being replaceable by the fluid from when it was queued in the insert check